changeset.go

```diff
@@ -129,65 +133,83 @@ func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, cha
 		return nil
 	}
 
+	// TODO Sergei: Provide instructions how to revert or delete a failed operation
 	if err != nil {
-		err = cs.Status(ctx, changesetNamespace, changesetName, id, DefaultRetryAttempts, DefaultRetryPeriod)
-		if err != nil {
-			return trace.Wrap(err, "Operation %s is not completed in changeset %s", id, changesetName)
-		}
+		return trace.Wrap(err)
+	}
+
+	ur := upsertResource{
+		resource: tr,
+		data:     data,
 	}
 
 	switch headers.TypeMeta.Kind {
 	case KindJob:
-		_, err = cs.upsertJob(ctx, tr, id, data)
+		_, err = cs.upsertJob(ctx, ur)
 	case KindDaemonSet:
-		_, err = cs.upsertDaemonSet(ctx, tr, id, data)
+		_, err = cs.upsertDaemonSet(ctx, ur)
 	case KindStatefulSet:
-		_, err = cs.upsertStatefulSet(ctx, tr, id, data)
+		_, err = cs.upsertStatefulSet(ctx, ur)
 	case KindReplicationController:
-		_, err = cs.upsertRC(ctx, tr, id, data)
+		_, err = cs.upsertRC(ctx, ur)
 	case KindDeployment:
-		_, err = cs.upsertDeployment(ctx, tr, id, data)
+		_, err = cs.upsertDeployment(ctx, ur)
 	case KindService:
-		_, err = cs.upsertService(ctx, tr, id, data)
+		_, err = cs.upsertService(ctx, ur)
 	case KindServiceAccount:
-		_, err = cs.upsertServiceAccount(ctx, tr, id, data)
+		_, err = cs.upsertServiceAccount(ctx, ur)
 	case KindConfigMap:
-		_, err = cs.upsertConfigMap(ctx, tr, id, data)
+		_, err = cs.upsertConfigMap(ctx, ur)
 	case KindSecret:
-		_, err = cs.upsertSecret(ctx, tr, id, data)
+		_, err = cs.upsertSecret(ctx, ur)
 	case KindRole:
-		_, err = cs.upsertRole(ctx, tr, id, data)
+		_, err = cs.upsertRole(ctx, ur)
 	case KindClusterRole:
-		_, err = cs.upsertClusterRole(ctx, tr, id, data)
+		_, err = cs.upsertClusterRole(ctx, ur)
 	case KindRoleBinding:
-		_, err = cs.upsertRoleBinding(ctx, tr, id, data)
+		_, err = cs.upsertRoleBinding(ctx, ur)
 	case KindClusterRoleBinding:
-		_, err = cs.upsertClusterRoleBinding(ctx, tr, id, data)
+		_, err = cs.upsertClusterRoleBinding(ctx, ur)
 	case KindPodSecurityPolicy:
-		_, err = cs.upsertPodSecurityPolicy(ctx, tr, id, data)
+		_, err = cs.upsertPodSecurityPolicy(ctx, ur)
 	default:
 		return trace.BadParameter("unsupported resource type %v", headers.TypeMeta.Kind)
 	}
 	return err
 }
 
-// isOpCompleted checks whether operation is completed in the context of a given changeset
-func (cs *Changeset) isOpCompleted(ctx context.Context, tr *ChangesetResource, id string) (bool, error) {
+// isOperationCompleted checks whether operation is completed in the context of a given changeset
+func (cs *Changeset) isOperationCompleted(ctx context.Context, tr *ChangesetResource, data []byte) (bool, error) {
+	headers, err := ParseResourceHeader(bytes.NewReader(data))
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
 	for _, op := range tr.Spec.Items {
-		if op.ID == id {
+		info, err := GetOperationInfo(op)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+
+		if info.To != nil &&
```

Comment:
```go
if info.equals(headers) ...
```
nit: Why is the variable called `headers` - does it describe multiple resources?